src/json.h

```diff
@@ -0,0 +1,52 @@
+/* ****************************************************************************
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ *
+ *
+ * This file is part of Microsoft R Host.
+ *
+ * Microsoft R Host is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Microsoft R Host is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Microsoft R Host.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * ***************************************************************************/
+
+#pragma once
+#include "stdafx.h"
+#include "util.h"
+#include "Rapi.h"
+
+namespace rhost {
+    namespace json {
+        // Produces JSON from an R object according to the following mappings:
+        //
+        // NULL, NA, empty vector -> null
+        // TRUE -> true
+        // FALSE -> false
+        // Vector of a single non-NA integer or double -> numeric literal
+        // Vector of a single non-NA string -> string literal
+        // List with all elements unnamed -> array (recursively)
+        // List with all elements named, or environment -> object (recursively)
+        //
+        // If any element of a list or environment is NA, that element is skipped.
```

Comment:
It is a shortcut to simplify building up large data structures in R code. Basically, it lets you preallocate a list for N elements, but then make "holes" in it as you scan your data and decide that some elements are not applicable. It also allows the convenient `NA_if_error` function to be used without any additional checks - you can see that a lot in `eval.R`.

(Keep in mind that this is not meant to be a general serialization solution for arbitrary R data structures, hence various constraints that it places on input. If you are passing a data structure directly from some R library to this serializer, you need to know very well what exactly the shape of that data is, otherwise you'll likely run afoul of other rules like no multiple-element vectrors, anyway.)

Having said that, I do somewhat dislike the use of `NA` specifically as sentinel. It is true that it's much more likely to appear in valid data, and it can be convenient to have it represented. I would like to replace that sentinel with an empty vector - `c()` - instead. However, I wanted to keep the behavior exactly the same as the old R serializer for now, since existing code in debugger relies on it, and that includes `NA` handling. Changes to improve the mapping can come later.